python/ray/air/execution/resources/fixed.py

```diff
@@ -0,0 +1,149 @@
+_DIGITS = 100000
+
+
+def _sum_bundle_resources(bundles: List[Dict[str, float]]) -> Dict[str, float]:
+    all_resources = {}
+    for resources in bundles:
+        for k, v in resources.items():
+            all_resources[k] = all_resources.get(k, 0) + v
+
+    return all_resources
+
+
+@DeveloperAPI
+@dataclass
+class FixedAllocatedResource(AllocatedResource):
+    bundles: List[Dict[str, float]]
+
+    def annotate_remote_objects(
+        self, objects: List[Type]
+    ) -> List[Union[ray.ObjectRef, ray.actor.ActorHandle]]:
+        # If we have an empty head, we schedule the first object (the "head") with
+        # empty resources.
+        if self.resource_request.head_bundle_is_empty:
+            bundles = [{}] + self.resource_request.bundles
+        else:
+            bundles = self.resource_request.bundles
+
+        if len(objects) > len(bundles):
+            raise RuntimeError(
+                f"The number of objects to annotate ({len(objects)}) cannot "
+                f"exceed the number of available bundles ({len(bundles)})."
+            )
+
+        annotated = []
+        for i, (obj, bundle) in enumerate(zip(objects, bundles)):
+            bundle = bundle.copy()
+            num_cpus = bundle.pop("CPU", 0)
+            num_gpus = bundle.pop("GPU", 0)
+            memory = bundle.pop("memory", 0.0)
+
+            annotated.append(
+                obj.options(
+                    num_cpus=num_cpus,
+                    num_gpus=num_gpus,
+                    memory=memory,
+                    resources=bundle,
+                )
+            )
+        return annotated
+
+
+@DeveloperAPI
+class FixedResourceManager(ResourceManager):
+    """Fixed budget based resource manager.
+
+    This resource manager keeps track of a fixed set of resources. When resources
+    are acquired, they are subtracted from the budget. When resources are freed,
+    they are added back to the budget.
+
+    The resource manager still requires resources to be requested before they become
+    available. However, because the resource requests are virtual, this will not
+    trigger autoscaling.
+
+    Additionally, resources are not reserved on request, only on acquisition. Thus,
+    acquiring a resource can change the availability of other requests. Note that
+    this behavior may be changed in future implementations.
+
+    Args:
+        total_resources: Budget of resources to manage. Defaults to all available
+            resources in the current task or all cluster resources (if outside a task).
+
+    """
+
+    _resource_cls: AllocatedResource = FixedAllocatedResource
+
+    def __init__(self, total_resources: Optional[Dict[str, float]] = None):
+        if not total_resources:
+            rtc = ray.get_runtime_context()
+            if rtc.worker.mode in {None, SCRIPT_MODE, LOCAL_MODE}:
+                total_resources = ray.available_resources()
+            else:
+                total_resources = rtc.get_assigned_resources()
```

Comment:
Would it make sense to always use this logic to fetch the current state of the world, rather than managing it in this class?

Reply:
This (optional) flexibility allows us to restrict resource usage (e.g. for parallel execution) and to trigger autoscaling, so we should keep it

Reply:
Oh actually I was thinking _about_ the autoscaler case, in which you might need to call `total_resources = ray.available_resources()` again some time after `__init__` (perhaps in `_available_resources`) in order to get an up to date view of available resources.

Reply:
Hmm, I'm not sure if this applicable here. Assuming we don't pass `total_resources` manually, so we take the current state of the world. With this resource manager, we will never allocate more resources than the current state of the world has. This means autoscaling won't be triggered (at least not by this resource manager).

If however we set `total_resources` to something higher than the current state of the world, autoscaling will be triggered (because we eagerly release resources), but we don't need to update the state as we have a fixed set of resources.

Thus, assuming no multi-tenancy, I don't think we need to update the state. Let me know if I'm missing something